Kentico.Kontent.Delivery.Caching/DeliveryCacheManager.cs

```diff
@@ -0,0 +1,161 @@
+
+namespace Kentico.Kontent.Delivery.Caching
+{
+    /// <summary>
+    /// Cache responses against the Kentico Kontent Delivery API.
+    /// </summary>
+    public sealed class DeliveryCacheManager : IDeliveryCacheManager
+    {
+        private readonly IMemoryCache _memoryCache;
+        private readonly DeliveryCacheOptions _cacheOptions;
+        private readonly ConcurrentDictionary<string, SemaphoreSlim> _createLocks = new ConcurrentDictionary<string, SemaphoreSlim>();
+        private readonly ConcurrentDictionary<string, object> _dependencyLocks = new ConcurrentDictionary<string, object>();
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="DeliveryCacheManager"/>
+        /// </summary>
+        /// <param name="memoryCache">An instance of an object that represent memory cache</param>
+        /// <param name="cacheOptions">The settings of the cache</param>
+        public DeliveryCacheManager(IMemoryCache memoryCache, IOptions<DeliveryCacheOptions> cacheOptions)
+        {
+            _memoryCache = memoryCache;
+            _cacheOptions = cacheOptions.Value ?? new DeliveryCacheOptions();
+        }
+
+        /// <summary>
+        /// Returns or Adds data to the cache
+        /// </summary>
+        /// <typeparam name="T">A generic type</typeparam>
+        /// <param name="key">A cache key</param>
+        /// <param name="valueFactory">A factory which returns a data</param>
+        /// <param name="shouldCache"></param>
+        /// <param name="dependenciesFactory"></param>
+        /// <returns>The data of generic type</returns>
+        public async Task<T> GetOrAddAsync<T>(string key, Func<Task<T>> valueFactory, Func<T, bool> shouldCache = null, Func<T, IEnumerable<string>> dependenciesFactory = null)
+        {
+            if (await TryGetAsync(key, out T entry))
+            {
+                return entry;
+            }
+
+            var entryLock = _createLocks.GetOrAdd(key, _ => new SemaphoreSlim(1, 1));
+            try
+            {
+                await entryLock.WaitAsync();
+
+                if (await TryGetAsync(key, out entry))
+                {
+                    return entry;
+                }
+
+                var value = await valueFactory();
+
+                // Decide if the value should be cached based on the response
+                if (shouldCache != null && !shouldCache(value))
+                {
+                    return value;
+                }
+
+                // Set different timeout for stale content
+                var valueCacheOptions = new MemoryCacheEntryOptions();
+                if (value is AbstractResponse ar && ar.HasStaleContent)
+                {
+                    valueCacheOptions.SetAbsoluteExpiration(_cacheOptions.StaleContentExpiration);
+                }
+                else
+                {
+                    valueCacheOptions.SetSlidingExpiration(_cacheOptions.DefaultExpiration);
+                }
```

Comment:
the original implementation of the time-bound cache manager always set absolute expiration. but I think this is fine.